OSDOCS-8373: Migrate "Getting Started with Rosa - Deploy the Cluster (Detailed CLI)" (#68159)
openshift/openshift-docs: cloud_experts_tutorials/cloud-experts-getting-started/cloud-experts-getting-started-deploying/cloud-experts-getting-started-detailed-cli-guide.adoc

* *`manual`:* Role and policies are created and saved in the current directory. You must manually run the provided commands as the next step. This option allows you to review the policy and roles before creating them.
* *`auto`:* Roles and policies are created and applied automatically using the current AWS account.

[TIP]
====
You can use either deployment method for this workshop. The `auto` mode is faster and has less steps.
====

== Deployment workflow
The overall deployment workflow follows these steps:

. `rosa create account-roles` - This is executed only _once_ for each account. Once created, the account roles do *not* need to be created again for more clusters of the same y-stream version.
. `rosa create cluster`
. `rosa create operator-roles` - For manual mode only.
. `rosa create oidc-provider` - For manual mode only.

For each additional cluster in the same account for the same y-stream version, only step 2 is needed for automatic mode. Steps 2 through 4 are needed for manual mode.

== Automatic mode
Use this method if you want the ROSA CLI to automate the creation of the roles and policies to create your cluster quickly.

=== Creating account roles

> If this is the _first time_ you are deploying ROSA in this account and you have _not_ yet created the account roles, then create the account-wide roles and policies, including Operator policies.

Author: I'm going to leave the Italics for now. They may go later, but I think they help when someone is skimming the tutorial.

I: Created role 'ManagedOpenShift-Support-Role' with ARN 'arn:aws:iam::000000000000:role/ManagedOpenShift-Support-Role'
I: Created role 'ManagedOpenShift-Installer-Role' with ARN 'arn:aws:iam::000000000000:role/ManagedOpenShift-Installer-Role'
I: Created policy with ARN 'arn:aws:iam::000000000000:policy/ManagedOpenShift-openshift-machine-api-aws-cloud-credentials'
I: Created policy with ARN 'arn:aws:iam::000000000000:policy/ManagedOpenShift-openshift-cloud-credential-operator-cloud-crede'
I: Created policy with ARN 'arn:aws:iam::000000000000:policy/ManagedOpenShift-openshift-image-registry-installer-cloud-creden'
I: Created policy with ARN 'arn:aws:iam::000000000000:policy/ManagedOpenShift-openshift-ingress-operator-cloud-credentials'
I: Created policy with ARN 'arn:aws:iam::000000000000:policy/ManagedOpenShift-openshift-cluster-csi-drivers-ebs-cloud-credent'
I: To create a cluster with these roles, run the following command:
    rosa create cluster --sts
----

=== Creating a cluster
Run the following command to create a cluster with all the default options:
[source,terminal]
----

> rosa create cluster --cluster-name <cluster-name> --sts --mode auto --yes

Author: There is a previous tutorial in this "workshop/tutorial" series that explains all of that.

+
Run the following command to create the needed files and output the additional commands:
+
[source,terminal]
----
rosa create account-roles --mode manual
----
+
.Example output
[source,terminal]
----
I: All policy files saved to the current directory
I: Run the following commands to create the account roles and policies:
aws iam create-role \
--role-name ManagedOpenShift-Worker-Role \
--assume-role-policy-document file://sts_instance_worker_trust_policy.json \
--tags Key=rosa_openshift_version,Value=4.8 Key=rosa_role_prefix,Value=ManagedOpenShift Key=rosa_role_type,Value=instance_worker
aws iam put-role-policy \
--role-name ManagedOpenShift-Worker-Role \
--policy-name ManagedOpenShift-Worker-Role-Policy \
--policy-document file://sts_instance_worker_permission_policy.json
----

> . Check the contents of your current directory to see the new files. We will use the `aws` CLI to create each of these objects.

Reviewer: Should this include "We will" or would just "Use (at beginning of sentence) suffice? I suppose another CLI tool could be used, so that may not work, but not sure if we should use the phrase "We will use"

+
.Example output
[source,terminal]
----
$ ls
openshift_cloud_credential_operator_cloud_credential_operator_iam_ro_creds_policy.json
sts_instance_controlplane_permission_policy.json
openshift_cluster_csi_drivers_ebs_cloud_credentials_policy.json        sts_instance_controlplane_trust_policy.json
openshift_image_registry_installer_cloud_credentials_policy.json          sts_instance_worker_permission_policy.json
openshift_ingress_operator_cloud_credentials_policy.json                 sts_instance_worker_trust_policy.json
openshift_machine_api_aws_cloud_credentials_policy.json                   sts_support_permission_policy.json
sts_installer_permission_policy.json                                      sts_support_trust_policy.json
sts_installer_trust_policy.json
----

> . *Optional:* Open the files to review what you will create. For example if we open the `sts_installer_permission_policy.json` we can see:

Reviewer: ```suggestion
. *Optional:* Open the files to review what you will create. For example, if we open the `sts_installer_permission_policy.json` we can see:
```

Reviewer: Also not a big fan of using "we" in this steps, but not sure if needs to be corrected in these tutorials?

+
.Example output
+
[source,terminal]
----
$ cat sts_installer_permission_policy.json
        {
        "Version": "2012-10-17",
        "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "autoscaling:DescribeAutoScalingGroups",
                "ec2:AllocateAddress",
                "ec2:AssociateAddress",
                "ec2:AssociateDhcpOptions",
                "ec2:AssociateRouteTable",
                "ec2:AttachInternetGateway",
                "ec2:AttachNetworkInterface",
                "ec2:AuthorizeSecurityGroupEgress",
                "ec2:AuthorizeSecurityGroupIngress",
                [...]
----
+

> You can also see the contents in the xref:../../../rosa_architecture/rosa-sts-about-iam-resources.adoc#rosa-sts-account-wide-roles-and-policies-creation-methods_rosa-sts-about-iam-resources[ROSA documentation].

Reviewer: Wording for link may need to be changed?

. Execute the `aws` commands listed in step 1. You can copy and paste as long as you are in the same directory as the JSON files you created.

=== Creating a cluster
. After the `aws` commands are executed successfully, run the following command to begin ROSA cluster creation in interactive mode:
+
[source,terminal]
----
rosa create cluster --interactive --sts
----
+
See the xref:../../../rosa_install_access_delete_clusters/rosa-sts-interactive-mode-reference.adoc#rosa-sts-interactive-mode-reference[ROSA documentation] for a description of the fields.

> . For the purpose of this tutorial, select the following values:

Reviewer: ```suggestion
. For the purpose of this tutorial, copy and then input the following values:
```

- arn:aws:iam::000000000000:role/my-rosa-cluster-openshift-machine-api-aws-cloud-credentials
- arn:aws:iam::000000000000:role/my-rosa-cluster-openshift-cloud-credential-operator-cloud-credentia
- arn:aws:iam::000000000000:role/my-rosa-cluster-openshift-image-registry-installer-cloud-credential
State:                      waiting (Waiting for OIDC configuration)
Private:                    No
Created:                    Jul  1 2022 22:13:50 UTC
Details Page:               https://console.redhat.com/openshift/details/s/2BMQm8xz8Hq5yEN000000000000
OIDC Endpoint URL:          https://rh-oidc.s3.us-east-1.amazonaws.com/1t6i760dbum4mqltqh6o000000000000
I: Run the following commands to continue the cluster creation:
rosa create operator-roles --cluster my-rosa-cluster
rosa create oidc-provider --cluster my-rosa-cluster
I: To determine when your cluster is Ready, run 'rosa describe cluster -c my-rosa-cluster'.
I: To watch your cluster installation logs, run 'rosa logs install -c my-rosa-cluster --watch'.
----
+
[NOTE]
====
The cluster state will remain as “waiting” until the next two steps are completed.
====

=== Creating Operator roles

> . At the end of the output in the above step, we are told the next commands to run. These roles need to be created _once_ for _each_ cluster. To create the roles run the following command:

Reviewer: Same comment about "we" and use of italics.

+
[source,terminal]
----
rosa create operator-roles --mode manual --cluster <cluster-name>
----
+
.Example output
+
[source,terminal]
----
I: Run the following commands to create the operator roles:
    aws iam create-role \
        --role-name my-rosa-cluster-openshift-image-registry-installer-cloud-credentials \
        --assume-role-policy-document file://operator_image_registry_installer_cloud_credentials_policy.json \
        --tags Key=rosa_cluster_id,Value=1mkesci269png3tck000000000000000 Key=rosa_openshift_version,Value=4.8 Key=rosa_role_prefix,Value= Key=operator_namespace,Value=openshift-image-registry Key=operator_name,Value=installer-cloud-credentials

    aws iam attach-role-policy \
        --role-name my-rosa-cluster-openshift-image-registry-installer-cloud-credentials \
        --policy-arn arn:aws:iam::000000000000:policy/ManagedOpenShift-openshift-image-registry-installer-cloud-creden
    [...]
----

. Run each of the `aws` commands.

=== Creating the OIDC provider

> . Run the following to create the OIDC provider:

Reviewer: ```suggestion
. Run the following command to create the OIDC provider:
```

+
[source,terminal]
----
rosa create oidc-provider --mode manual --cluster <cluster-name>
----

> . This displays the `aws` commands that you need to run. Run the commands like the below example:

Reviewer: ```suggestion
. This displays the `aws` commands that you need to run. As an example:
```

+
[source,terminal]
----
I: Run the following commands to create the OIDC provider:
$ aws iam create-open-id-connect-provider \
--url https://rh-oidc.s3.us-east-1.amazonaws.com/1mkesci269png3tckknhh0rfs2da5fj9 \
--client-id-list openshift sts.amazonaws.com \
--thumbprint-list a9d53002e97e00e043244f3d170d000000000000

$ aws iam create-open-id-connect-provider \
--url https://rh-oidc.s3.us-east-1.amazonaws.com/1mkesci269png3tckknhh0rfs2da5fj9 \
--client-id-list openshift sts.amazonaws.com \
--thumbprint-list a9d53002e97e00e043244f3d170d000000000000
----

. Your cluster will now continue the installation process.  

=== Checking the installation status
. Run one of the following commands to check the status of your cluster:
+
* For a detailed view of the status, run:
+
[source,terminal]
----
rosa describe cluster --cluster <cluster-name>
----
* For an abridged view of the status, run:
+
[source,terminal]
----
rosa list clusters
----

The cluster state will change from “waiting” to “installing” to "ready". This will take about 40 minutes.

> . Once the state changes to “ready” your cluster is installed.

Reviewer: Not sure this should be numbered? Made similar comment earlier in the review....

== Obtaining the console URL
* To obtain the console URL, run the following commmand:
+
[source,terminal]
----
rosa describe cluster -c <cluster-name> | grep Console
----

> The cluster has now been successfully deployed. In the next step we will create an admin user to be able to use the cluster immediately.

Reviewer: Is there a next step?
